[REVERTED] Fixes Iterable#toString forcing elements

eed3si9n opened the pull request:
Fixes https://github.com/scala/bug/issues/11785

This changes the implementation of Iterable#toString to not force the evaluation of its elements unless it extends StrictOptimizedIterableOps.

### note

I considered pushing this into `StrictOptimizedIterableOps` trait as originally suggested in https://github.com/scala/bug/issues/11785, but adding new methods to the trait would introduce bincompat breakage.

som-snytt commented:
That's interesting. I expected the fix in `ScalaRunTime.stringOf`.

joshlemer commented on src/library/scala/collection/Iterable.scala:
```diff
@@ -69,12 +69,8 @@ trait Iterable[+A] extends IterableOnce[A]
-    *            by all elements separated by commas and enclosed in parentheses.
     */
-  override def toString = mkString(className + "(", ", ", ")")
+  override def toString: String  = className + "(<not computed>)"
```
I don't think we can really say `<not computed>` as the default, it's more like in a state of `<unsure if computed or not>`. Maybe the Iterable should either not implement a toString, or should leave it as similar to `Iterator`'s implementation? (`iterator.toString` is  `<iterator>`).

eed3si9n replied:
So `<iterable>`?

joshlemer replied:
That's correct 😄

som-snytt replied:
If you wanted to be fancy, you could report `"of $knownSize elements"` if non-negative.

I guess `ScalaRunTime.stringOf` only custom prints `StrictOptimizedIterable`, to pretty print `List[Array[String]]` for example. Otherwise `toString`.

eed3si9n replied:
Changed to `<iterable>` in https://github.com/scala/scala/pull/8554/commits/e0c5211af040b41a829d156149288e7fab5ac76c

szeiger commented:
I don't like the idea of overriding it to use the strict version in `Seq`, `Set` and `Map`. While lazy sets and maps are questionable, it is completely expected to have a lazy `Seq` so this change doesn't fix anything unless you explicitly extend `Iterable` instead of `Seq`. The differentiating factor needs to be `StrictOptimizedIterableOps`, and if you cannot put it directly into that trait you need to check for it in `Iterable`.

eed3si9n commented:
@szeiger

>  if you cannot put it directly into that trait you need to check for it in Iterable.

That sounds like a good idea.

eed3si9n committed:
Fixes Iterable#toString forcing elements

Fixes scala/bug#11785

This changes the implementation of Iterable#toString to not force the evaluation of its elements unless it extends StrictOptimizedIterableOps.

eed3si9n commented on src/library/scala/collection/Iterable.scala:
```diff
@@ -68,12 +68,12 @@ trait Iterable[+A] extends IterableOnce[A]
+    this match {
+      case _: StrictOptimizedIterableOps[_, _, _] => mkString(className + "(", ", ", ")")
+      case _                                      => className + "(<iterable>)"
+    }
```
@szeiger Does this look ok?

joshlemer replied:
I feel like this is not really ok. According to the name of these `StrictOptimized______Ops`, they are meant to not change behaviour at all, just to provide performance optimizations for strict collections. But for instance, users who define their own collections may forget to inherit from `StrictOptimized________Ops`. Currently,  the only consequence is that they may not get the absolute best performance, but with these changes, they'll actually have broken `toString` methods.

som-snytt replied:
In `stringOf`, it's trying to accommodate custom iterables that have their own `toString`.

eed3si9n committed:
Un-deprecate default floating point Orderings, and change to migration

Fixes https://github.com/scala/bug/issues/11844
Ref https://github.com/scala/bug/issues/10511
Ref https://github.com/scala/scala/pull/6410
Ref https://github.com/scala/scala/pull/76

This change the deprecation of `DeprecatedDoubleOrdering` to a migration warning instead to avoid `List(1.0, -1.0).sorted` giving deprecation warning.

This also provides some documentation on the ordering instances in Scaladoc.

lrytz commented:
There seems to be some unintended change, for example

```
!!    1 - run/Meter.scala                           [output differs]
% diff /home/jenkins/workspace/scala-2.13.x-validate-main@2/test/files/run/Meter.check /home/jenkins/workspace/scala-2.13.x-validate-main@2/test/files/run/Meter-run.log
@@ -10,7 +10,7 @@ x == 1: false
 x == y: true
 a == b: true
 testing native arrays
-Array(1.0m, 2.0m)
+Array(<iterable>)
 1.0m
 >>>1.0m<<< 1.0m
 >>>2.0m<<< 2.0m
```

https://scala-ci.typesafe.com/job/scala-2.13.x-validate-main/10971/testReport/junit/test.files/run_/Meter_scala/

eed3si9n commented:
@lrytz Yea. Looks like there's a fake array called prettyArray in https://github.com/scala/scala/blob/2.13.x/src/partest/scala/tools/partest/Util.scala that needs to be updated.

diesalbla committed:
Extract the call to updateAttachment as a lambda.

In some profiles, we noticed that the call to `clone.updateAttachment`
was generating up to 10 MiB of lambda objects.
Since this seems to be reentrant, we can save those allocations by
using a single Function1 object with a mutable state.

diesalbla committed:
Use alternative approach: no foreach.

It turns out that we may be able to avoid the foreach and the lambda
altogether, and instead use a cloneAttachments function.

retronym committed:
Merge pull request #8718 from diesalbla/symbols_update_attachment_lambda

Symbols: share the set of Attachments

diesalbla committed:
Contexts: reduce allocations in imports lookup

Add a fast path for when an import doesn't supply the
given name. We can avoid a Tuple2 allocation here and
and lambda or two.

SethTisue committed:
Merge pull request #8721 from eed3si9n/wip/migrate-ordering

Un-deprecate default floating point Orderings; issue migration warning instead under -Xmigration

retronym committed:
Merge pull request #8728 from retronym/review/8727

Contexts: reduce allocations in imports lookup

lrytz committed:
Merge pull request #8554 from eed3si9n/wip/iterabletostring

Fixes Iterable#toString forcing elements

lrytz commented:
Hmm, I clicked "merge", and the commit went in (https://github.com/scala/scala/commits/2.13.x), but this PR didn't close. Maybe related to the commit sha?

<img width="90" alt="image" src="https://user-images.githubusercontent.com/119636/74847576-a3374500-5332-11ea-91e3-f4b438463712.png">

😱 💀 😵 ☠️ ⚰️

SethTisue commented:
https://www.githubstatus.com is reporting trouble currently

joshlemer commented:
Reposting my comment here for visibility:
 
According to the name of these StrictOptimized______Ops traits, they are meant to not change behaviour at all, rather just to provide performance optimizations for strict collections. But for instance, users who define their own collections may forget to inherit from StrictOptimized________Ops. Currently, the only consequence is that they may not get the absolute best performance, but with these changes, they'll actually have broken toString methods. 

Are we certain that we're ok with breaking any custom strict collections which didn't happen to inherit from StrictOptimized_______Ops?

eed3si9n commented:
@joshlemer 

> Are we certain that we're ok with breaking any custom strict collections which didn't happen to inherit from StrictOptimized_______Ops?

Just saw this comment now. The original bug https://github.com/scala/bug/issues/11785 said the problem is that `toString` forces evaluation. What do you suggest that won't break custom strict collection that also solves https://github.com/scala/bug/issues/11785?

lrytz commented:
(GitHub didn't recover the state of this PR to "merged". Tried some random git push, which closed this PR, but also made its diff look wrong. Maybe @eed3si9n you can still `push -f` the `dead51ba01` revision to your branch to fix what shows up in the diff here? In any case, no big deal.)

eed3si9n commented:
I just push -f'ed:

<img width="1019" alt="Commits_·_eed3si9n_scala" src="https://user-images.githubusercontent.com/184683/74943858-14263d80-53c3-11ea-9fd0-f66237509d8a.png">

but the PR continues to show 9 commits 🤷‍♂️

sjrd commented:
Once a PR is closed, it is frozen. Changes to the branch (fast-forward, force push, removal) do not affect the PR anymore.

Moreover, once the branch is not in sync with the frozen state of the PR anymore, the PR cannot be reopened :p

SethTisue commented:
caused this community build failure: https://scala-ci.typesafe.com/job/scala-2.13.x-integrate-community-build/3216/artifact/logs/scala-collection-contrib-build.log

```
[scala-collection-contrib] [error] Test scala.collection.MultiDictTest.testToString failed: org.junit.ComparisonFailure: expected:<MultiDict([<iterable>])> but was:<MultiDict([1 -> 1])>, took 0.002 sec
[scala-collection-contrib] [error]     at scala.collection.MultiDictTest.runForFactory$1(MultiDictTest.scala:107)
[scala-collection-contrib] [error]     at scala.collection.MultiDictTest.testToString(MultiDictTest.scala:117)
[scala-collection-contrib] [error]     ...
```

looks like it might be an example of exactly what Josh was afraid would happen?

som-snytt commented:
https://gitter.im/scala/contributors?at=5e4f4864ff00c664eed6f80b

"The community build couldn't exist without volunteers willing to break it."

SethTisue commented:
@joshlemer @szeiger I fooled a bit with trying to alter `MultiDict` and `MultiSet` to fix this, but it wasn't as simple as replacing `IterableOps` with `StrictOptimizedIterableOps`

note that in the above error log, the JUnit assertion in question was written backwards (with actual an expected in the wrong order), so the error message is backwards, too.

SethTisue commented:
reverting, for 2.13.2 anyway, as per https://github.com/scala/scala-dev/issues/671
